### DISCUSSION TITLE
How do I set up a big data set of similar structures, so that I can set a specific key:value if I know the parent array index / object id?

### DISCUSSION THREAD

**Discussion by Splendorr** [pmndrs/jotai#1380] [Q&A]
Hi everybody! I'm getting the hang of Jotai, used it for lots of simple data in my current app. Now I'm trying to work with a larger data set, and I'm having trouble. I can't tell if I'm over- or under-engineering, or just "engineering" up the wrong tree. 😄 

In short, I'm creating a set of TV Channels, 1-999, which each need to store various data points (status in the example can equal "Vacant", "Occupied", or a few other values. There will be other values in the same object, like `{... callsign: "ESPN2"}`). The real dataset is more complex, which is why I need to generate the full set and populate it with API results. Then, a new set of channel details (or, importantly, data related to one specific channel) can be retrieved, and I need to update the dataset.

Here's a sandbox where things half-work. I was building against (I think) Jotai 1.4. The createAllChannels() function doesn't work in Jotai 1.8.1 to actually instantiate all the channels: https://codesandbox.io/s/laughing-visvesvaraya-pq3vht?file=/src/App.js

Expected behavior would be that on page load, I could enter any value 1–999 in the text input, see a Status string of "Occupied," and use the button to persist a change to that channel number's Status. Right now I see "Occupied" on load, then don't see it for any values until I manually set them.

Later in the real app, I'm successfully creating a derived atom to filter what's visible based on a few criteria. But I've omitted that from the example to keep it simple. I've also successfully displayed all the values in my real app.

@dai-shi suggested when I posed this in a comment that I might not need (or understand) atomFamily. It felt like the right idea for a dataset consisting of identically-structured items, but isn't getting the results I was hoping for.

Given that it's a sequential set of channel numbers, I know I could probably just use an array and refer to the index. I foresee future datasets that aren't as straightforward. I guess I'm not clear on how to create a large array of similar objects, and then most efficiently read/write to a specific one. `splitAtom` is good for mapping to display a bunch of items, but doesn't seem right for working with a specific atom. `selectAtom` could derive a bunch of atoms from an original array, but I don't want to declare 999 separate atoms. And again I'm not sure how to cleanly change just the `status` of channel 183.

One final note: I'm not performing these functions from inside of channel-specific components. So I can't just map over the dataset and have each visible component change itself. I need to make channel-specific updates from the parent component that controls the API requests by pointing directly into the data.

So, optimally I'd be able to create a single atom to contain all of this information. Knowing a channel number and new status or callsign, I'd be able to set channel 88's callsign to "ESPN" without affecting the other data. Basically if I had a regular array of objects, `allChannels[88].callsign = "ESPN"`... but in the proper Jotai way to get all the benefits of minimal re-renders and derived dependencies.

Thanks for your time, I know I'm probably wrong about multiple parts of this! I've re and re-read the docs and searched the discussions, and can't figure out "the best way" to handle this.

**Comment by dai-shi** [comment]
I'm not sure if I follow your description well, but one thing is atomFamily is just a cache, so if you want to initialize it, you want it in advance: https://codesandbox.io/s/cocky-cdn-msxrtl

That's probably not what you want, because there's no way to list all atoms in an atom family.
I would just create atoms in an array in such a case: https://codesandbox.io/s/proud-cherry-ovv7lj

If there seems to be some misunderstanding of mine, I hope you would re-describe a smaller problem. (I know you already tried so.)

**Reply by Splendorr** [reply]
... I'm laughing at myself because "an array of atoms" is so much simpler. Definitely in the phase of learning a new tool, and trying to apply it to everything. I was trying make the *container* an atom, too. But for this purpose just putting atoms in a regular array seems simple and perfect.

I'll test it out today and report back!
